File: flutter/feedback_service.dart
import 'dart:convert';
import 'dart:io';

import 'package:http/http.dart' as http;
import 'package:package_info_plus/package_info_plus.dart';

import 'feedback_config.dart';

// FeedbackService
// Sends feedback via Telegram, Discord, Slack, or Proxy depending on FeedbackConfig.channel.
// Uses the `http` package -- the only external dependency.

class FeedbackResult {
  final bool success;
  final String? error;

  const FeedbackResult.success() : success = true, error = null;
  const FeedbackResult.failure(this.error) : success = false;
}

class FeedbackService {
  static Future<FeedbackResult> send({
    required String text,
    http.Client? client,
  }) async {
    final trimmed = text.trim();
    if (trimmed.isEmpty) {
      return const FeedbackResult.failure('Feedback text cannot be empty.');
    }

    final Uri url;
    final String body;

    switch (FeedbackConfig.channel) {
      case FeedbackChannel.telegram:
        final message = await buildMessage(text: trimmed);
        final token = FeedbackConfig.botToken;
        url = Uri.parse('https://api.telegram.org/bot$token/sendMessage');
        body = jsonEncode({
          'chat_id': FeedbackConfig.chatID,
          'text': message,
          'parse_mode': 'Markdown',
        });
        break;
      case FeedbackChannel.discord:
        final message = await buildMessage(text: trimmed);
        url = Uri.parse(FeedbackConfig.webhookURL);
        body = jsonEncode({'content': message});
        break;
      case FeedbackChannel.slack:
        final message = await buildMessage(text: trimmed);
        url = Uri.parse(FeedbackConfig.webhookURL);
        body = jsonEncode({'text': message});
        break;
      case FeedbackChannel.proxy:
        url = Uri.parse(FeedbackConfig.proxyURL);
        final packageInfo = await PackageInfo.fromPlatform();
        final platformName = Platform.isIOS ? 'iOS' : 'Android';
        final osVersion = Platform.operatingSystemVersion;
        body = jsonEncode({
          'app_name': FeedbackConfig.appName,
          'app_version': packageInfo.version,
          'platform': '$platformName $osVersion',
          'feedback': trimmed,
        });
        break;
    }

    try {
      final httpClient = client ?? http.Client();
      final response = await httpClient.post(
        url,
        headers: {'Content-Type': 'application/json'},
        body: body,
      );

      if (client == null) httpClient.close();

      // Discord returns 204 No Content on success
      if (response.statusCode >= 200 && response.statusCode < 300) {
        return const FeedbackResult.success();
      } else {
        return FeedbackResult.failure(
          'API error (${response.statusCode}): ${response.body}',
        );
      }
    } catch (e) {
      return FeedbackResult.failure('Network error: ${e.toString()}');
    }
  }

  static Future<String> buildMessage({
    required String text,
  }) async {
    final appName = FeedbackConfig.appName;

    String appVersion;
    try {
      final packageInfo = await PackageInfo.fromPlatform();
      appVersion = packageInfo.version;
    } catch (_) {
      appVersion = 'Unknown';
    }

    final platformName = Platform.isIOS ? 'iOS' : 'Android';
    final osVersion = Platform.operatingSystemVersion;
    final platformEmoji = Platform.isIOS ? '\u{1F34E}' : '\u{1F916}';

    final now = DateTime.now().toUtc();
    final timestamp =
        '${now.year}-${now.month.toString().padLeft(2, '0')}-${now.day.toString().padLeft(2, '0')} '
        '${now.hour.toString().padLeft(2, '0')}:${now.minute.toString().padLeft(2, '0')}';

    return '\u{1F4F1} App: $appName\n'
        '\u{1F4E6} Version: $appVersion\n'
        '$platformEmoji Platform: $platformName $osVersion\n'
        '\u{1F554} Time: $timestamp UTC\n'
        '\n'
        '\u{1F4AC} Feedback:\n'
        '$text';
  }
}

File: react-native/FeedbackService.ts
import { Platform } from "react-native";
import * as Application from "expo-application";
import { FeedbackConfig } from "./FeedbackConfig";

// FeedbackService
// Sends feedback via Telegram, Discord, Slack, or Proxy depending on FeedbackConfig.channel.
// Uses native `fetch` -- no third-party networking dependencies.

export type FeedbackResult =
  | { success: true }
  | { success: false; error: string };

export async function sendFeedback(
  text: string,
  category?: string
): Promise<FeedbackResult> {
  const trimmed = text.trim();
  if (!trimmed) {
    return { success: false, error: "Feedback text cannot be empty." };
  }

  let url: string;
  let body: string;

  switch (FeedbackConfig.channel) {
    case "telegram": {
      const message = buildMessage(trimmed, category);
      const token = FeedbackConfig.botToken;
      url = `https://api.telegram.org/bot${token}/sendMessage`;
      body = JSON.stringify({
        chat_id: FeedbackConfig.chatID,
        text: message,
        parse_mode: "Markdown",
      });
      break;
    }
    case "discord": {
      const message = buildMessage(trimmed, category);
      url = FeedbackConfig.webhookURL;
      body = JSON.stringify({ content: message });
      break;
    }
    case "slack": {
      const message = buildMessage(trimmed, category);
      url = FeedbackConfig.webhookURL;
      body = JSON.stringify({ text: message });
      break;
    }
    case "proxy": {
      const appVersion =
        Application.nativeApplicationVersion ?? "Unknown";
      const platformName = Platform.OS === "ios" ? "iOS" : "Android";
      const osVersion = Platform.Version;

      url = FeedbackConfig.proxyURL;
      body = JSON.stringify({
        app_name: FeedbackConfig.appName,
        app_version: appVersion,
        platform: `${platformName} ${osVersion}`,
        feedback: trimmed,
        ...(category && category.trim() ? { category } : {}),
      });
      break;
    }
  }

  try {
    const response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body,
    });

    // Discord returns 204 No Content on success
    if (response.ok) {
      return { success: true };
    }

    const responseBody = await response.text();
    return {
      success: false,
      error: `API error (${response.status}): ${responseBody}`,
    };
  } catch (e) {
    const errorMessage =
      e instanceof Error ? e.message : "Unknown error";
    return { success: false, error: `Network error: ${errorMessage}` };
  }
}

export function buildMessage(text: string, category?: string): string {
  const appName = FeedbackConfig.appName;
  const appVersion =
    Application.nativeApplicationVersion ?? "Unknown";

  const platformName = Platform.OS === "ios" ? "iOS" : "Android";
  const osVersion = Platform.Version;
  const platformEmoji = Platform.OS === "ios" ? "\u{1F34E}" : "\u{1F916}";

  const now = new Date();
  const pad = (n: number) => n.toString().padStart(2, "0");
  const timestamp = `${now.getUTCFullYear()}-${pad(now.getUTCMonth() + 1)}-${pad(now.getUTCDate())} ${pad(now.getUTCHours())}:${pad(now.getUTCMinutes())}`;

  const lines = [
    `\u{1F4F1} App: ${appName}`,
    `\u{1F4E6} Version: ${appVersion}`,
    `${platformEmoji} Platform: ${platformName} ${osVersion}`,
    `\u{1F554} Time: ${timestamp} UTC`,
  ];

  if (category && category.trim()) {
    lines.push(`\u{1F3F7}\u{FE0F} Type: ${category}`);
  }

  lines.push("", `\u{1F4AC} Feedback:`, text);

  return lines.join("\n");
}
